src/platform/ESP32/nimble/BLEManagerImpl.cpp

```diff
@@ -0,0 +1,926 @@
+            ExitNow();
+        }
+
+        ExitNow();
+    }
+
+    // If the application has enabled CHIPoBLE and BLE advertising...
+    if (mServiceMode == ConnectivityManager::kCHIPoBLEServiceMode_Enabled &&
+        GetFlag(mFlags, kFlag_AdvertisingEnabled)
+#if CHIP_DEVICE_CONFIG_CHIPOBLE_SINGLE_CONNECTION
+        // and no connections are active...
+        && (_NumConnections() == 0)
+#endif
+    )
+    {
+        // Start/re-start advertising if not already advertising, or if the advertising state of the
+        // ESP BLE layer needs to be refreshed.
+        if (!GetFlag(mFlags, kFlag_Advertising) || GetFlag(mFlags, kFlag_AdvertisingRefreshNeeded))
+        {
+            // Configure advertising data if it hasn't been done yet.  This is an asynchronous step which
+            // must complete before advertising can be started.  When that happens, this method will
+            // be called again, and execution will proceed to the code below.
+            if (!GetFlag(mFlags, kFlag_AdvertisingConfigured))
+            {
+                err = ConfigureAdvertisingData();
+                ExitNow();
+            }
+
+            // Start advertising.  This is also an asynchronous step.
+            err = StartAdvertising();
+            ExitNow();
+        }
+    }
+
+    // Otherwise stop advertising if needed...
+    else
+    {
+        if (GetFlag(mFlags, kFlag_Advertising))
+        {
+            err = ble_gap_adv_stop();
+            if (err != CHIP_NO_ERROR)
+            {
+                ChipLogError(DeviceLayer, "ble_gap_adv_stop() failed: %s", ErrorStr(err));
+                ExitNow();
+            }
+
+            SetFlag(mFlags, kFlag_ControlOpInProgress);
+
+            ExitNow();
+        }
+    }
+
+    // Stop the CHIPoBLE GATT service if needed.
+    if (mServiceMode != ConnectivityManager::kCHIPoBLEServiceMode_Enabled && GetFlag(mFlags, kFlag_GATTServiceStarted))
+    {
+        // TODO: what to do about existing connections??
+    }
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(DeviceLayer, "Disabling CHIPoBLE service due to error: %s", ErrorStr(err));
+        mServiceMode = ConnectivityManager::kCHIPoBLEServiceMode_Disabled;
+    }
+}
+
+CHIP_ERROR BLEManagerImpl::InitESPBleLayer(void)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(!GetFlag(mFlags, kFlag_ESPBLELayerInitialized), /* */);
+
+    err = esp_nimble_hci_and_controller_init();
+
+    nimble_port_init();
+
+    SetFlag(mFlags, kFlag_ESPBLELayerInitialized);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR BLEManagerImpl::ConfigureAdvertisingData(void)
+{
+    CHIP_ERROR err;
+    struct ble_hs_adv_fields fields = { 0 }, scan_rsp = { 0 };
+
+    ESP32ChipServiceData chipServiceData;
+
+    // If a custom device name has not been specified, generate a CHIP-standard name based on the
+    // bottom digits of the Chip device id.
+    const uint16_t discriminator = 0x0F00;
```

Comment:
Can we retrieve the discriminator from Config and use digits instead of hex ? 
```
    // If a custom device name has not been specified, generate a CHIP-standard name based on the
    // discriminator value
    uint32_t discriminator;
    SuccessOrExit(err = ConfigurationMgr().GetSetupDiscriminator(discriminator));

    if (!GetFlag(mFlags, kFlag_UseCustomDeviceName))
    {
        snprintf(mDeviceName, sizeof(mDeviceName), "%s%04u", CHIP_DEVICE_CONFIG_BLE_DEVICE_NAME_PREFIX, discriminator);
        mDeviceName[kMaxDeviceNameLength] = 0;
    }
```

Reply:
@vivien-apple yes the suggested changes are included